Risk management and RoSH summary pages pull data from OASys

bryangaledxw opened the pull request:
# Changes in this PR

* Updates to our OASys import code, copied from CAS1
* Change to our RoSH level page - showing a warning if risk data could not be read from OASys
* Changes to our risk management plan page - getting questions and pre-filled answers from OASys, or falling back to JSON-driven questions, rather than using a static set of questions
* A versioning pattern, used on our risk management plan page

## Screenshots of UI changes

### Before

#### RoSH level page
![rosh before](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/7e2f9d56-8196-4562-bbc5-85b13bdc6274)

#### Risk management page
![rmp before](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/ae04b51b-d8d3-47d9-b83a-dad776518dda)

### After

#### RoSH level page
![rosh after](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/bda30064-fb97-44be-b0ff-2ab4a5eba31f)

#### Risk management page
![rmp after](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/1a523cbe-5ac4-46b2-ab28-1ae592bdcafb)

#### Risk management page errors
![rmp after error](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/dc0195e5-b4e6-46f1-869f-4dd3aa02d2d9)

#### Risk management page CYA
![cya after](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/7c3fdbbc-8f76-47ec-ab6c-a673a98e1650)


# Release checklist

[Release process documentation](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/edit-v2/4247847062?draftShareId=a1c360ab-bd31-4db1-aae3-7cc002761de9).

## Pre-merge checklist

- [ ] Are any changes required to dependent services for this change to work?
  (eg. CAS API)
- [ ] Have they been released to production already?

## Post-merge checklist

- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to test
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to preprod
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to prod

<!-- Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible. You can monitor deployment failures in CircleCI
itself and application errors are found in
[Sentry](https://ministryofjustice.sentry.io/issues/?project=4504129156218880&referrer=sidebar&statsPeriod=24h).
Both events should be automatically sent to our [Slack
channel](https://mojdt.slack.com/archives/C048BJS7S2F). -->

bryangaledxw committed:
Update oasysImportUtils from CAS1

Ahead of updating our risk management plan page to import data from OASys, we update oasysImportUtils and related files to match the latest versions from CAS1

bryangaledxw committed:
Update risk management plan Apply page

We update this page to import questions and pre-fill their answers from OASys, using getOasysSections(). In the case where OASys is not available, this will fall back to using the data in oasysStubs.json.

This replaces a previous version of this page that used a fixed set of questions.

bryangaledxw committed:
Reduce Cypress numTestsKeptInMemory

We reduce the number of tests Cypress keeps in memory during integration tests, to reduce the frequency of crashes when running locally

bryangaledxw committed:
Update getOasysSections() to only persist known data

We update getOasysSections() to only persist known data. This ensure that even if the API sends additional data, the data we store for a given Apply page will match the expected schema

bryangaledxw committed:
Update getOasysSections() to use questionKeyFromNumber()

We add questionKeyFromNumber() and questionNumberFromKey() helper functions, and use them within getOasysSections(). This prevents any question keys that can be parsed as a number, which would result in an Apply page getting an array back from a view, rather than a map

bryangaledxw committed:
Update getOasysSections() to swallow any errors from PersonService

Out of caution, we swallow any errors from PersonService, to allow any Apply pages using getOasysSections() to be completed, even if attempting to retrieve OASys data is causing API errors

bryangaledxw committed:
Update RoSH level Apply page

We update this page to display further text if RoSH information could not be retrieved

bryangaledxw committed:
Update dates persisted by getOasysSections()

We record the date OASys was last updated on the Apply page body, and let that take precedence over any new date if it already exists. Previously, if a user completed a page, then OASys data was updated, and then the user returned to the page, we might display that OASys has been recently updated, when in fact we're still using old OASys data that has persisted in the page body.

We effectively move from possibly marking stale data as up-to-date, to possibly marking up-to-date data as stale (with the former case occurring if the user leaves the imported data in place, and the latter case occurring if the user manually deletes the imported data).

We also now record the date we imported data from OASys for display by an Apply page.

bryangaledxw committed:
Update forPagesInTask()

We update forPagesInTask() to throw an error if it encounters a page that has no data, or a page that has errors. Our motivation for this is to prevent silent failures when our Apply form changes while a user has part way completed the form. This could otherwise lead to a submission that doesn't match what the user sees on the check your answers page

bryangaledxw commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Risk management and RoSH summary pages pull data from OASys](https://trello.com/c/XSZMaoNK/1397-risk-management-and-rosh-summary-pages-pull-data-from-oasys)

bryangaledxw committed:
Fix getTaskStatus()

We fix how getTaskStatus() checks if an Apply page has an errors, which previously could give false negatives. A ticket has been added to our backlog to address wider issues with the tests around this method, that allowed tests to pass despite this problem (https://trello.com/c/Dp1Uqj7W/1426-address-problems-with-gettaskstatus-tests)

bryangaledxw committed:
Add version checking to risk management plan Apply page

The previous version of this page has been deployed and is being used by users. Once this updated version is deployed, we want to ensure that any users that have not yet submitted the form are forced to revisit this page. We do this by adding a hidden "version" field to the Nunjucks template, and checking it in our errors() method.

As this version field will not be present in data generated by the previous version of this page, errors() will return an error until the user revisits this page. And thanks to our changes in pagesForTask(), they will encounter an error if they attempt to view check your answers, or submit the application before doing so. And thanks to our fix in getTaskStatus(), they will not encounter an error in practice, unless they are very unlucky with the timing of a deploy. Instead, the task list will show the task containing this page as incomplete until they revisit it.

bryangaledxw commented:
@tahb A few things I want to flag:

* When we can't communicate with OASys, we fall back to using the questions in [oasysStubs.json](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/blob/main/server/data/stubs/oasysStubs.json), as per CAS1. I've not delved into whether those are definitely the right questions for CAS3

* Given that getOasysSections() is necessarily called both when the page is first displayed and when the user clicks save, and given that data is effectively co-owned (imported from OASys, but editable by the user), it's quite hard to reason whether we're doing the right thing when it comes to displaying the "last updated" date. I *think* my change in my "Update dates persisted by getOasysSections()" is an improvement, but I think there are still some failure cases

* The final commit I think is a decent approach for page versioning in general, in the absence of time for a more sophisticated system

bryangaledxw commented:
I've also extracted this ticket - https://trello.com/c/GKSfwwdF/1430-fix-question-sorting-on-risk-management-plan-apply-screen

tahb commented:
I had started digging into the risk management plan questions, [discovering there would be at most 8](https://github.com/ministryofjustice/hmpps-probation-integration-services/blob/5a10c0aa4da90bd77183391921f9c927783453d7/projects/approved-premises-and-oasys/src/main/kotlin/uk/gov/justice/digital/hmpps/model/RiskManagementPlanDetails.kt#L33). I'll go and check with Selina and Barry to figure out if there's any reason not to consider all 8.

tahb commented on server/views/applications/pages/assess-placement-risks-and-needs/placement-considerations/risk-management-plan.njk:
```diff
@@ -16,6 +16,8 @@
   <div class="govuk-grid-row">
     <div class="govuk-grid-column-two-thirds">
 
+      <input type="hidden" name="version" value="{{ page.latestVersion }}"/>
```
Very nice approach. We can layer this approach into each question if/when we need to.

tahb reviewed (approved):
This looks good to me. I've got [a thread going with other folks](https://mojdt.slack.com/archives/C03GF35FY7P/p1691078467053259) to be clear about handling 8 (instead of 4) risk management questions + more if that changes in future.

I think the oasys import date choices are also reasonable and we'll see what users make of them since this has come from the central team.

tahb commented:
@bryangaledxw [Selina has suggested that the 4 new sections of the RMP aren't useful for our users](https://www.figma.com/file/lgKst3F6PAgmIFnEtKdfLe?node-id=2478:134759&mode=design#511327526). How much work would it be to put an allow list around the 4 questions we care about so that we forget/ignore the remaining 4 for either manual and automatic input?

bryangaledxw commented:
@tahb I don't think huge, if you're happy for me to spend more time on this. On Selina's point about sorting, I pulled out a ticket for that, as I didn't feel I had time to address that - https://trello.com/c/GKSfwwdF/1430-fix-question-sorting-on-risk-management-plan-apply-screen

tahb commented:
@bryangaledxw Agree on sorting, they should still be able to succeed in spite of that.

I've talked the idea of removing the 4 new questions through with Selina some more and I think we have to do default back to the original 4 questions please. 

The original 4 were on the old CPP referral forms and have been tested. The new questions are untested and may cause problems.

Referrals are still disabled in prod so you have the option of merging this in as is and adding a follow up PR. That may also give us an easier way to revert and add the questions all back in should we learn that HPTs really need them after all.

bryangaledxw committed:
Filter risk management questions

We filter the risk management questions pulled from OASys to a known subset that are relevant to CAS3 users. We also update the factory used when stubbing risk data so we have questions available for local testing

bryangaledxw commented:
@tahb Question filtering is in if you want to take another look. Relevant commit is "Filter risk management questions"

tahb reviewed (approved):
LGTM

(I do wonder if the RM question numbers will change one day. As we should fail safely it's not too big a worry)

bryangaledxw commented:
@tahb Yeah, I think ideal world this filtering would be API side. Checking the RM numbers seemed safer than checking the labels, and in any case, if things change, the E2Es will fail so we'll know about it
